Add StackFrame.locals() method

brenns10 opened the pull request:
I know you mentioned in #134 that listing variables in a stack frame would be something that is made part of the Scopes API (#123). That makes sense, but given that the Scopes API may be a long time coming, and it will need to cover a lot of complexity, I wanted to share this small implementation of this idea. No expectations here (even of review, let alone merging :smile: ). I'm starting to play around with debugger code built on top of this, so I'm personally finding it useful.

osandov commented on libdrgn/drgn.h.in:
```diff
@@ -2785,6 +2785,20 @@ struct drgn_error *drgn_stack_frame_symbol(struct drgn_stack_trace *trace,
+ * @return @c NULL on success, non-@c NULL on error
+ */
+struct drgn_error *
+drgn_stack_frame_list_locals(struct drgn_stack_trace *trace, size_t frame_i,
+			     const char ***names_ret, size_t *count_ret);
```
I'd prefer to call this `drgn_stack_frame_locals()`.

osandov commented on libdrgn/dwarf_info.c:
```diff
@@ -5357,6 +5357,38 @@ static struct drgn_error *find_dwarf_enumerator(Dwarf_Die *enumeration_type,
+			switch (dwarf_tag(&die)) {
+			case DW_TAG_variable:
+			case DW_TAG_formal_parameter:
+				if (count == alloc) {
+					alloc = max(alloc * 2, 16UL);
+					names = realloc(names, alloc * sizeof(char *));
+				}
```
You can use `vector.h` instead of open-coding this.

brenns10 replied:
I never know what use case is "too small" for `vector.h`, since it feels silly cluttering up with all those functions. On the other hand, I guess they're all static and so they have every opportunity to get compiled away if unused, or inlined as necessary. I'll be sure to use `vector.h` next time I see myself reallocating an array :P

osandov commented on libdrgn/dwarf_info.c:
```diff
@@ -5357,6 +5357,38 @@ static struct drgn_error *find_dwarf_enumerator(Dwarf_Die *enumeration_type,
+					alloc = max(alloc * 2, 16UL);
+					names = realloc(names, alloc * sizeof(char *));
+				}
+				names[count++] = dwarf_diename(&die);
```
`dwarf_diename()` can return `NULL`, in which case you should skip the DIE. (I ran into this recently; see 330c71b5b59906c1336df9f2de19257c20f749ab).

osandov commented on _drgn.pyi:
```diff
@@ -1738,6 +1738,13 @@ class StackFrame:
+        """
+        Get a list of all local variable names in local scopes for this stack
+        frame. Not all names may have present values, but they can be used with
+        __getitem__() to check.
```
Those less familiar with the Python data model would probably recognize `[]` rather than `__getitem__`. We do it like this elsewhere in the documentation:

```suggestion
        the :meth:`[] <.__getitem__>` operator to check.
```

osandov reviewed (changes requested):
You've convinced me that this would be valuable to have in the short-term without the full-fledged scopes API. We can always redo this on top of the scopes API later if needed.

I was considering whether we should have a separate `StackFrame.arguments()` method to get just the function arguments, but that might get too messy with the possibility of local variables shadowing function arguments. That part should probably wait for the scopes API.

A few comments inline.

brenns10 commented:
I rebased and added all your suggestions. I've been looking into adding tests for stack unwinding, so that I can add some tests for the .locals() method. I started with the `tests/sample.coredump.zst` file, but couldn't find corresponding DWARF info. So I whipped up a quick program to create my own core dump and got the debuginfo for libc with debuginfod. Getting the backtrace works with GDB, but I only get the first frames with drgn. I'm not sure if the issue is on my side or drgn. [Here's](https://gist.github.com/brenns10/cc57bcb4154759c6edba5e803c8e87f7) the quick program I had made to produce the core dump. I ended up getting errors like this:
```
$ python -m drgn -c dumpfile -s coredump -s libc.so.debuginfo 
drgn 0.0.21+5.g25e4615.dirty (using Python 3.10.6, elfutils 0.186, with libkdumpfile)
For help, type help(drgn).
>>> import drgn
>>> from drgn import NULL, Object, cast, container_of, execscript, offsetof, reinterpret, sizeof
>>> from drgn.helpers.common import *
>>> t = list(prog.threads())
>>> t[0].stack_trace()
#0  __clock_nanosleep (../sysdeps/unix/sysv/linux/clock_nanosleep.c:78:7)
>>> t[1].stack_trace()
#0  __clock_nanosleep (../sysdeps/unix/sysv/linux/clock_nanosleep.c:78:7)
```
Compared to GDB's core stack traces like this:
```
(gdb) symbol-file libc.so.debuginfo 
Load new symbol table from "libc.so.debuginfo"? (y or n) y
Reading symbols from libc.so.debuginfo...
(gdb) symbol-file coredump
Load new symbol table from "coredump"? (y or n) y
Reading symbols from coredump...
(gdb) bt
#0  0x00007f35acde6868 in __GI___clock_nanosleep (clock_id=clock_id@entry=0, flags=flags@entry=0, req=req@entry=0x7f35ab4f9d40, rem=rem@entry=0x7f35ab4f9d40) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:78
#1  0x00007f35acdeb6e7 in __GI___nanosleep (req=req@entry=0x7f35ab4f9d40, rem=rem@entry=0x7f35ab4f9d40) at ../sysdeps/unix/sysv/linux/nanosleep.c:25
#2  0x00007f35acdeb61e in __sleep (seconds=0) at ../sysdeps/posix/sleep.c:55
#3  0x00005603e5e71380 in last (arg_a=123, arg_b=456, arg_c=789, tid=83410) at coredump.c:44
#4  0x00005603e5e71334 in baz (arg_a=123, arg_b=456, arg_c=789, tid=83410) at coredump.c:40
#5  0x00005603e5e712ad in bar (arg_a=123, arg_b=456, arg_c=789, tid=83410) at coredump.c:39
#6  0x00005603e5e71226 in foo (arg_a=123, arg_b=456, arg_c=789, tid=83410) at coredump.c:38
#7  0x00005603e5e713cd in thread_run (arg=0x0) at coredump.c:51
#8  0x00007f35acd95b43 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#9  0x00007f35ace27a00 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
(gdb) q
```
I did some printf debugging and am 95% certain it has to do with debuginfo - I noticed it was failing in the call to `drgn_module_find_cfi()`. I can email you all the relevant files if you want to reproduce it, though I don't know if it will matter post debuginfo discovery. Of course we can move this to an issue if you think this is a bug worth tracking.
---

In any case, I'll look at the linux_kernel stack trace tests and see what I can put in there to help test the locals() method. If the above issue turns out to be a genuine bug, then that's probably an even better argument for including real tests for userspace stack traces, so I'd prefer to have tests in both places :)

osandov commented:
The CFI thing is probably the known bug where drgn tries to get `.eh_frame` from the wrong file. I was actually just working on a fix for that. I was originally going to fix that as part of the module API, but I figured out how to do it independently. I just pushed it to https://github.com/osandov/drgn/tree/eh-frame. I still had a couple of things I wanted to tweak (and I need to write a commit message for it), but it should be good enough for you to test.

There is another issue where `-s` doesn't work for user space. That's another thing I was going to fix as part of the module API, but I have an idea for how to fix it without.

If you can test it in the existing Linux kernel tests, that's probably the easiest way forward for now.

brenns10 commented:
You're totally right, sorry for the added noise! I just applied this on top of your branch and the stack traces (and `.locals()` method) worked great.

I'll aim to test this change in the linux_kernel tests for now. However I might create a branch for some tests for userspace stack_trace tests which we could merge later. It would be cool to have both when possible. While I don't do a ton of userspace work with drgn it's nice to have the option and have test coverage there too.

brenns10 commented:
I really like the idea behind creating a userspace core dump (with debuginfo) and using that for testing userspace debugging, like the Thread API has. Of course to get core dumps consistently working you need control over `kernel.core_pattern`. Do you think it would be a bad idea to put some sort of userspace testing via a simple C program that we compile for each architecture (like the kernel module helper), which can do an `abort()`, and then we can examine the core dump? That way we could test multiple architectures. I know this is off topic for this PR, but it's something I could try to add for a branch regarding stack trace userspace testing.

brenns10 commented:
Ok, this should be ready for another look!

- Addressed your code review comments
- Added a stack frame test in the linux_kernel helper code. I use the test kmod to create a kthread which just sits in uninterruptible state, and then I look up the locals from that frame.

osandov commented on tests/linux_kernel/test_stack_trace.py:
```diff
@@ -94,3 +99,17 @@ def test_stack__repr_pretty_(self):
+        for frame in stack_trace:
+            if frame.symbol().name == "drgn_kthread_fn":
+                self.assertSetEqual(
+                    set(["arg", "a", "b", "c"]),
```
Use a set literal.

```suggestion
                    {"arg", "a", "b", "c"},
```

brenns10 replied:
I truly wrote too much python 2.6-compatible code in a past life...

osandov commented on tests/linux_kernel/kmod/drgn_test.c:
```diff
@@ -385,6 +386,45 @@ static int drgn_test_slab_init(void)
+
+static struct task_struct *drgn_kthread;
+
+static int __attribute__((optimize("O0"))) drgn_kthread_fn(void *arg)
```
Lol at `O0`. I guess I can't think of a better way to do it.

brenns10 replied:
Haha yeah, I'm sorry for that, but it did seem to do the trick.

I guess I don't actually test whether the contents of the variables are present, so the attribute could be dropped. DWARF should have the names of the variables and parameters regardless. I had it around in the userspace version of my test so that I could also lookup the values. But I don't know if it matters much either way.

osandov reviewed (changes requested):
Gave it a closer look this time, so there's some stuff here I missed in the last iteration.

osandov commented:
> I really like the idea behind creating a userspace core dump (with debuginfo) and using that for testing userspace debugging, like the Thread API has. Of course to get core dumps consistently working you need control over `kernel.core_pattern`. Do you think it would be a bad idea to put some sort of userspace testing into the vmtest system? We can set `kernel.core_pattern` in the VM because we have root. Then just compile a simple C program on each architecture (like the kernel module helper), which can do an `abort()`, and then we can examine the core dump? That way we could test multiple architectures. I know this is off topic for this PR, but it's something I could try to add for a branch regarding stack trace userspace testing.

That does seem like a useful integration test. Making it work everywhere might be tricky, since you'll need to get debug info for libc and such.

brenns10 commented on libdrgn/dwarf_info.c:
```diff
@@ -5353,6 +5353,43 @@ static struct drgn_error *find_dwarf_enumerator(Dwarf_Die *enumeration_type,
+	struct drgn_string_vector vec;
+
+	drgn_string_vector_init(&vec);
+	for (size_t scope = num_scopes - 1; scope > 0; scope--) {
```
Ah thank you, this was the part I was uncertain about when I wrote it. And come to think of it, I should have just looked into the definition of `drgn_stack_frame` and the solution would have jumped out at me. Hardcoding 0 was a serious bozo move :P

brenns10 commented on libdrgn/dwarf_info.c:
```diff
@@ -5353,6 +5353,43 @@ static struct drgn_error *find_dwarf_enumerator(Dwarf_Die *enumeration_type,
+		do {
+			switch (dwarf_tag(&die)) {
+			case DW_TAG_variable:
+			case DW_TAG_formal_parameter:
```
Got it, I've "pattern matched" my implementation based off of what I saw you do with `find_dwarf_enumerator()`, but please do make sure I'm not doing stupid here. I actually know nothing about libdw, and really should make it a priority to learn it!

brenns10 commented:
> you'll need to get debug info for libc and such.

I was hoping to use debuginfod to come to the rescue for this. I'll see if I can do it. I noticed that you pushed the eh_frame fix to main so I can give it a shot in the clear now!

---

I think this revision handles your concerns! Enjoy it Monday!

osandov commented on libdrgn/dwarf_info.c:
```diff
@@ -5336,6 +5336,77 @@ drgn_object_from_dwarf(struct drgn_debug_info *dbinfo,
+	int r = dwarf_child(enumeration_type, &child);
+	while (r == 0) {
+		if (dwarf_tag(&child) == DW_TAG_enumerator) {
+			const char *name = dwarf_diename(&child);
```
We need to check that `name` isn't `NULL` (which I realize is also a bug in `find_dwarf_enumerator()`).

osandov replied:
Fixed the `find_dwarf_enumerator()` version of this bug in b3a5051ff446a6e54aa217dda87685bd4a494f9c.

osandov commented on libdrgn/dwarf_info.c:
```diff
@@ -5336,6 +5336,77 @@ drgn_object_from_dwarf(struct drgn_debug_info *dbinfo,
+				bool enum_class;
+				if (dwarf_flag_integrate(&die, DW_AT_enum_class,
+							 &enum_class))
+					return drgn_error_libdw();
```
Needs to `goto out_err` instead of `return`.

brenns10 replied:
oof 😅

osandov commented on libdrgn/stack_trace.c:
```diff
@@ -377,6 +377,29 @@ drgn_stack_frame_symbol(struct drgn_stack_trace *trace, size_t frame,
+	struct drgn_stack_frame *frame = &trace->frames[frame_i];
+
+	if (frame->function_scope >= frame->num_scopes) {
+		names_ret = NULL;
+		count_ret = 0;
```
```suggestion
		*names_ret = NULL;
		*count_ret = 0;
```

brenns10 replied:
oof 😅

osandov reviewed (commented):
I caught a few last-minute subtle bugs, but I'll fix those on my end (along with some cosmetic changes) so I can merge this. I'm really excited for this!

brenns10 commented:
Ah nice catches, thanks for the careful review and quick update!

osandov commented:
Looks like I broke Packit; see #221. The rest passed, so I'll go ahead and merge this. Thanks again!
